File: src/main/java/toker/warbandscripts/panel/controller/rest/BoardController.java
package toker.warbandscripts.panel.controller.rest;

import com.fasterxml.jackson.annotation.JsonView;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PutMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RestController;
import toker.warbandscripts.panel.entity.NoticeBoard;
import toker.warbandscripts.panel.entity.NoticeBoardAccess;
import toker.warbandscripts.panel.service.BoardService;

import java.util.List;

@RestController
public class BoardController {

    @Autowired
    private BoardService boardService;

    public BoardController(BoardService boardService) {
        this.boardService = boardService;
    }

    @GetMapping("/api/board")
    @JsonView(NoticeBoard.View.None.class)
    public List<NoticeBoard> boards() {
        return boardService.getAllBoards();
    }

    @PutMapping("/api/player/boardAccess")
    @JsonView(NoticeBoardAccess.View.Board.class)
    public NoticeBoardAccess saveBoardAccess(@RequestBody NoticeBoardAccess boardAccess) {
        return boardService.saveBoardAccess(boardAccess);
    }
}

File: src/main/java/toker/panel/controller/rest/BoardController.kt
package toker.panel.controller.rest

import com.fasterxml.jackson.annotation.JsonView
import org.springframework.data.crossstore.ChangeSetPersister
import org.springframework.security.access.prepost.PreAuthorize
import org.springframework.web.bind.annotation.*
import toker.panel.bean.SelectedServerId
import toker.panel.entity.*
import toker.panel.entity.NoticeBoardAccess.View.Board
import toker.panel.entity.pk.DoorPK
import toker.panel.entity.pk.NoticeBoardPK
import toker.panel.repository.NoticeBoardRepository
import toker.panel.service.BoardService

@RestController
class BoardController(
    private val boardService: BoardService,
    private val boardRepository: NoticeBoardRepository
) {

    @GetMapping("/api/board")
    @JsonView(NoticeBoard.View.None::class)
    fun boards(): List<NoticeBoard> {
        return boardService.allBoards
    }

    @GetMapping("/api/board/{index}")
    @JsonView(NoticeBoard.View.Accesses::class)
    fun board(@PathVariable index: Int): NoticeBoard {
        return boardRepository.findOne { root, query, builder ->
            root.fetch(NoticeBoard_.accesses)
            query.distinct(true)
            builder.and(
                builder.equal(root.get(NoticeBoard_.index), index),
                builder.equal(root.get(NoticeBoard_.server).get(Server_.id), SelectedServerId),
            )
        }.orElseThrow { ChangeSetPersister.NotFoundException() }
    }

    @PutMapping("/api/board/{index}")
    @JsonView(NoticeBoard.View.None::class)
    fun board(@RequestBody door: NoticeBoard, @PathVariable index: Int): NoticeBoard {
        val newState = boardRepository.findById(NoticeBoardPK(index = index, server = SelectedServerId))
            .orElseThrow { ChangeSetPersister.NotFoundException() }
        newState.name = door.name
        return boardRepository.saveAndFlush(newState)
    }
}